### PR TITLE
deprecate fallback constructor

### LINKED ISSUE #15120: Falling back to Base.convert during construction does not work for custom types with a single field
When I run the following code

``` julia
type A
    v::Float64
end

type B
    v::Float64
end

Base.convert(::Type{A}, b::B) = A(b.v/2)
b = B(2)
A(b)
```

it fails with this error message because the implicit constructor is called.

``` julia
ERROR: MethodError: `convert` has no method matching convert(::Type{Float64}, ::A)
This may have arisen from a call to the constructor Float64(...),
since type constructors fall back to convert methods.
Closest candidates are:
  call{T}(::Type{T}, ::Any)
  convert(::Type{Float64}, ::Int8)
  convert(::Type{Float64}, ::Int16)
  ...
 in call at none:2
```

I expected this to work since the manual states:

> defining `Base.convert(::Type{T}, args...) = ...` automatically defines a constructor `T(args...) = ....`"

The following code on the other hand works:

``` julia
type A
    v::Float64
end

Base.convert(::Type{Float64}, a::A) = a.v/12
a = A(2)
Float64(a)
```

It seems that the new constructor is only generated for built-in types and not for custom types.

EDIT: Typo.

### PR DESCRIPTION
**Pull request by JeffBezanson**
This resolves #15120. The approach used here is to define explicit constructors for all types (as a bonus, it's nice that more constructors can be defined in `Core` where the types themselves are defined), and then add `convert` methods to indicate which constructors are safe for use in conversion. This also has the nice side effect of removing an unsightly warning from the system image build, since `Core.Inference` can work without the fallback constructor.

@nanosoldier `runbenchmarks(ALL, vs=":master")`

### CODE REVIEW

File: base/int.jl [hunk 1]
```diff
@@ -31,8 +31,8 @@ const BitUnsigned64T   = Union{Type{UInt8}, Type{UInt16}, Type{UInt32}, Type{UIn
 
 const BitIntegerType = Union{map(T->Type{T}, BitInteger_types)...}
 
-throw_inexacterror(f::Symbol, ::Type{T}, val) where T =
-    (@_noinline_meta; throw(InexactError(f, T, val)))
+# calls constructors defined in boot.jl
+convert(T::BitIntegerType, x::Union{BitInteger, Bool}) = T(x)
```
Comment (vtjnash) [inline base/int.jl]: in abstract_call_method, there's a heuristic that should be changed to reflect the new signature:
```julia
        # the construct-to-convert method is a bottleneck in inference,
        # so just assume that recursion will get prevented at some other point
       && !(method.sig == Tuple{Type, Any}))
```
that should address the inference failure in mmap

Comment (vtjnash) [inline base/int.jl]: I thought we suggested that this should be:
```julia
convert(::Type{I}, x) where {I <: Number} = I(x)
```

Comment (JeffBezanson) [inline base/int.jl]: Yes, it maybe should be eventually, I was just being conservative for now.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23273) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/a1453b07d9b3281e54cd6f688a7ac7d7a26429ff/3d74919_vs_0df3f7d/report.md). cc @ararslan

**Comment by JeffBezanson** [comment]
Ah, what happened here was that the type assertion in 
```
(::Type{T})(arg) where {T} = convert(T, arg)::T
```
was very helpful to inference in some cases. The new constructors are all inferable, but there are too many of them, so we hit the method match limit (#23210).

**Comment by vtjnash** [comment]
Can we just remove that limit?

**Comment by JeffBezanson** [comment]
That would be a good experiment. I guess we can rely on the property that worse type information will tend to make the inferred return type diverge to `Any` quickly, so we can still skip many methods. The main thing to do is to make `jl_matching_methods` return some kind of iterator, so we don't need to build a full list of matches when we're only going to look at 2 of them. (That would probably be a good change anyway.)

**Comment by JeffBezanson** [comment]
Here's a small wrinkle. It's pretty popular to define highly generic conversions like these:

```
convert(::Type{T}, x::TwicePrecision) where {T<:Number} = convert(T, x.hi + x.lo)

convert(::Type{T}, x::Enum{T2}) where {T<:Integer,T2<:Integer} = convert(T, bitcast(T2, x))
```

Do we want these to be constructors? We could certainly move to a world where things like this are defined as constructors by default, with only a relatively small number of `convert` methods. I think that's the plan, just want to spell it out clearly.

**Comment by vtjnash** [comment]
That second one should never have existed as a method of convert, so +1 to continuing with the plan.

**Comment by JeffBezanson** [comment]
I looked into the couple slowdowns here. I can reproduce the `revcomp` one locally, but the LLVM IR is identical. It might be due to dispatch table rearrangement; the benchmark involves a `Box`. (It's also slower on master than on 0.6.)

**Comment by vtjnash** [comment]
We should also merge https://github.com/JuliaLang/julia/pull/22300

**Comment by JeffBezanson** [comment]
Updated to change all `Number` conversions to constructors, and deprecate a few more conversions that seems sketchy to me.

**Comment by JeffBezanson** [comment]
This is now in pretty good shape.

The last remaining issue was the return type of e.g. `map(Integer, Any[])`. That used to give `Integer[]`, but here gave `Any[]` since we are, against our better judgment, using inference to determine this type :) To get around that, I restored a rule we once had (and which was not discussed much) to use `T` itself as the default_eltype in `map(T::Type, ...)`. The main problem with that is that `map(T, ...)` and `map(x->T(x), ...)` can give different results. But I'm not sure what else to do here --- our only other choices are to stop using inference for this, or decide we don't care about the type of `map(T, Any[])` and remove those tests.

**Comment by StefanKarpinski** [comment]
Could we just take `map(T::Type, v)` as meaning `map(x->T(x)::T, v)` and use the fact that we know the return type must be `T` to get the element type right here? I feel like that's not too far a stretch since it's kind of a special case of requiring that `T` actually construct `T` objects. (Perhaps this is exactly what you've done.)

**Comment by JeffBezanson** [comment]
Yes that's exactly what I did.

**Comment by JeffBezanson** [comment]
Now getting this:

```
julia> code_typed(+, (Rational{UInt128},Rational{UInt128}), optimize=false)
1-element Array{Any,1}:
 CodeInfo(:(begin
        nothing
        SSAValue(0) = (Base.divgcd)((Core.getfield)(x, :den)::UInt128, (Core.getfield)(y, :den)::UInt128)::Tuple{Any,Any}
...

julia> code_typed(Base.divgcd, (UInt128,UInt128), optimize=false)
1-element Array{Any,1}:
 CodeInfo(:(begin
        nothing
        g = (Base.gcd)(x, y)::UInt128
        #= line 24 =#
        return (Core.tuple)((Base.div)(x, g)::UInt128, (Base.div)(y, g)::UInt128)::Tuple{UInt128,UInt128}
    end)) => Tuple{UInt128,UInt128}
```

**Comment by JeffBezanson** [comment]
ENOMEM on freebsd. This has also been known to trigger what seems to be an inference issue (https://github.com/JuliaLang/julia/pull/23273#issuecomment-341800408), but that hasn't happened for about a month. Mergeable? @vtjnash What do you think?

**Comment by vtjnash** [comment]
That machine, IIRC, has more memory than any other buildbot (and the place where we could get ENOMEM from calling `uv_write` are allocating trivial amounts of memory). Perhaps something here is causing memory corruption or enabling it to remove an essential gc root?

**Comment by ararslan** [comment]
cc @iblis17

**Comment by iblis17** [comment]
hmm, let me try more builds

builds tracking list
- https://julia.iblis.cnmc.tw/#/builders/1/builds/4917
- https://julia.iblis.cnmc.tw/#/builders/1/builds/4925
- https://julia.iblis.cnmc.tw/#/builders/1/builds/4926
- https://julia.iblis.cnmc.tw/#/builders/1/builds/4927

**Comment by iblis17** [comment]
All following build pass
well, it's another heisenbug?

**Comment by iamed2** [comment]
Will this be happening for 1.0?

**Comment by StefanKarpinski** [comment]
Yes, it will. Jeff is on vacation for the week.

**Comment by JeffBezanson** [comment]
Getting this after rebase, if somebody wants to dig in:
```
julia: /home/jeff/src/julia/src/APInt-C.cpp:455: void LLVMTrunc(unsigned int, llvm::integerPart*, unsigned int, llvm::integerPart*): Assertion `inumbits > onumbits' failed.

signal (6): Aborted
in expression starting at loading.jl:335
raise at /build/buildd/eglibc-2.19/signal/../nptl/sysdeps/unix/sysv/linux/raise.c:56
abort at /build/buildd/eglibc-2.19/stdlib/abort.c:89
__assert_fail_base at /build/buildd/eglibc-2.19/assert/assert.c:92
__assert_fail at /build/buildd/eglibc-2.19/assert/assert.c:101
LLVMTrunc at /home/jeff/src/julia/src/APInt-C.cpp:455
jl_intrinsic_cvt at /home/jeff/src/julia/src/runtime_intrinsics.c:400
jl_call_fptr_internal at /home/jeff/src/julia/src/julia_internal.h:380 [inlined]
jl_call_method_internal at /home/jeff/src/julia/src/julia_internal.h:399 [inlined]
jl_apply_generic at /home/jeff/src/julia/src/gf.c:2011
jl_apply at /home/jeff/src/julia/src/julia.h:1474 [inlined]
jl_f__apply at /home/jeff/src/julia/src/builtins.c:556
builtin_tfunction at ./inference.jl:1774
```

**Comment by JeffBezanson** [comment]
Fixed that issue and rebased; let's see what happens now.

File: base/linalg/cholesky.jl [hunk 2]
```diff
@@ -349,30 +349,30 @@ function cholfact(x::Number, uplo::Symbol=:U)
...
+Factorization{T}(C::CholeskyPivoted) where {T} = CholeskyPivoted{T}(C)
 
-convert(::Type{AbstractMatrix}, C::Cholesky) = C.uplo == 'U' ? C[:U]'C[:U] : C[:L]*C[:L]'
-convert(::Type{AbstractArray}, C::Cholesky) = convert(AbstractMatrix, C)
-convert(::Type{Matrix}, C::Cholesky) = convert(Array, convert(AbstractArray, C))
-convert(::Type{Array}, C::Cholesky) = convert(Matrix, C)
+AbstractMatrix(C::Cholesky) = C.uplo == 'U' ? C[:U]'C[:U] : C[:L]*C[:L]'
+AbstractArray(C::Cholesky) = AbstractMatrix(C)
+Matrix(C::Cholesky) = Array(AbstractArray(C))
+Array(C::Cholesky) = Matrix(C)
```
Comment (Sacha0) [inline base/linalg/cholesky.jl]: Beautiful! :) All factorization->array convert methods become array constructors in this pull request I imagine?

**Comment by JeffBezanson** [comment]
This has gotten a bit out of hand with the large number of changes happening, but please bear with me. I think we really need this in 1.0 --- it can't be handled neatly by renaming or deprecations.

**Comment by StefanKarpinski** [comment]
100% agree; just take care of it when you're back!

**Comment by JeffBezanson** [comment]
Tests passing again! Anybody know if these CircleCI failures are new?

**Comment by ararslan** [comment]
Nope, not new.

**Comment by StefanKarpinski** [comment]
They're not new, but they're unrelated afaict. They're both remote worker exceptions and Linux builds pass on Travis, so I think you're in the clear here. Circle CI has been shot for the past week or more.

File: base/linalg/cholesky.jl (hunk 2 above)
Comment (JeffBezanson) [reply base/linalg/cholesky.jl]: Yes, I think so.

**Comment by StefanKarpinski** [comment]
@JeffBezanson, you should probably merge this when you're ready – it's passing CI and we don't want to let it get any new conflicts!

**Comment by JeffBezanson** [comment]
:tada:

**Pull request merged**